Src/Plugins/ValidatorBase.cs

```diff
@@ -41,19 +42,47 @@ protected ValidatorBase()
         /// </summary>
         protected IDictionary<string, Tuple<string, string>> FingerprintToResultCache { get; }
 
+        /// <summary>
+        /// Gets a cache of file + fingerprint combinations that have been
+        /// observed previously. Our scanner will only detect and validate
+        /// a unique fingerprint once per file. Many regexes will produce
+        /// multiple matches that resolve to the same unique credential in
+        /// a file (one of the perils of multiline regex matching). It is
+        /// possible that this cache may drop the location of a second
+        /// actual match that happens to be duplicated in the file. In
+        /// practice, we will not worry about this scenario: it will be
+        /// sufficient that we flag one instance of the unique secret.
+        /// </summary>
+        protected ISet<string> PerFileFingerprintCache { get; }
+
         public static string IsValidStatic(ValidatorBase validator,
                                            ref string matchedPattern,
                                            ref Dictionary<string, string> groups,
                                            ref string failureLevel,
                                            ref string fingerprint,
                                            ref string message)
         {
-            return
+            string state =
                 validator.IsValidStaticHelper(ref matchedPattern,
                                               ref groups,
                                               ref failureLevel,
                                               ref fingerprint,
                                               ref message);
+
+            if (state != nameof(ValidationState.NoMatch))
+            {
+                string scanTarget = groups["scanTargetFullPath"];
```

Comment:
>groups [](start = 36, length = 6)

This convention is how we flow the scan target path to the plug-ins. Now, we could have created a new parameter instead. But this identifier is a well-known thing in the scanner format strings, so I thought it was fine to preserve this well-known value in the groups object.